python/ray/autoscaler/cli_logger.py

```diff
@@ -0,0 +1,426 @@
+
+    def newline(self):
+        self._print('')
+
+    def _print(self, msg, linefeed=True):
+        if self.old_style:
+            return
+
+        if self.strip:
+            msg = _strip_codes(msg)
+
+        rendered_message = "  " * self.indent_level + msg
+
+        if not linefeed:
+            sys.stdout.write(rendered_message)
+            sys.stdout.flush()
+            return
+
+        print(rendered_message)
+
+    def indented(self, cls=False):
+        cli_logger = self
+        class IndentedContextManager():
+            def __enter__(self):
+                cli_logger.indent_level += 1
+
+            def __exit__(self, type, value, tb):
+                cli_logger.indent_level -= 1
+
+        if cls:
+            # fixme: this does not work :()
+            return IndentedContextManager
+        return IndentedContextManager()
+
+    def timed(self, msg, *args, **kwargs):
+        self.print(msg, *args, **kwargs)
+
+        # todo: implement timer
+        cli_logger = self
+        class TimedContextManager():
+            def __enter__(self):
+                cli_logger.indent_level += 1
+
+            def __exit__(self, type, value, tb):
+                cli_logger.indent_level -= 1
+
+        return TimedContextManager()
+
+    def group(self, msg, *args, **kwargs):
+        self._print(_format_msg(cf.cornflowerBlue(msg), *args, **kwargs))
+
+        return self.indented()
+
+    def verbatim_error_ctx(self, msg, *args, **kwargs):
+        cli_logger = self
+        class VerbatimErorContextManager():
+            def __enter__(self):
+                cli_logger.error(cf.bold("!!! ")+msg, *args, **kwargs)
+
+            def __exit__(self, type, value, tb):
+                cli_logger.error(cf.bold("!!!"))
+
+        return VerbatimErorContextManager()
+
+    def keyvalue(self, key, msg, *args, **kwargs):
+        self._print(
+            cf.cyan(key)+": "+_format_msg(cf.bold(msg), *args, **kwargs))
+
+    def verbose(self, msg, *args, **kwargs):
+        if self.verbosity > 0:
+            self.print(msg, *args, **kwargs)
+
+    def verbose_error(self, msg, *args, **kwargs):
+        if self.verbosity > 0:
+            self.error(msg, *args, **kwargs)
+
+    def very_verbose(self, msg, *args, **kwargs):
+        if self.verbosity > 1:
+            self.print(msg, *args, **kwargs)
+
+    def success(self, msg, *args, **kwargs):
+        self._print(_format_msg(cf.green(msg), *args, **kwargs))
+
+    def warning(self, msg, *args, **kwargs):
+        self._print(_format_msg(cf.yellow(msg), *args, **kwargs))
+
+    def error(self, msg, *args, **kwargs):
+        self._print(_format_msg(cf.red(msg), *args, **kwargs))
+
+    def print(self, msg, *args, **kwargs):
+        self._print(_format_msg(msg, *args, **kwargs))
+
+    def abort(self, msg=None, *args, **kwargs):
+        if msg is not None:
+            self.error(msg, *args, **kwargs)
+
+        sys.exit(1)
```

Comment:
I think we should avoid this; the product also wraps the autoscaler and suddenly exiting the process can be unexpected/undesirable.

Can we instead raise a ClickException, which hides stack traces for ray cli users but also allows other libraries to function properly?

Reply:
a good CLI never has an unhandled exception
if a library runs into this code, they must have deliberately set `old_style` to `false`, so it's their problem